packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -14,71 +14,249 @@
+    headers,
+  };
+}
+
+// Converts for example
+// from: https://bitbucket.org/orgname/reponame/src/master/file.yaml
+// to:   https://api.bitbucket.org/2.0/repositories/orgname/reponame/src/master/file.yaml
+export function getApiUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+    if (
+      !owner ||
+      !name ||
+      (filepathtype !== 'browse' &&
+        filepathtype !== 'raw' &&
+        filepathtype !== 'src')
+    ) {
+      throw new Error('Invalid Bitbucket URL or file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+
+    if (provider.host === 'bitbucket.org') {
+      if (!ref) {
+        throw new Error('Invalid Bitbucket URL or file path');
+      }
+      return new URL(
+        `${provider.apiBaseUrl}/repositories/${owner}/${name}/src/${ref}/${pathWithoutSlash}`,
+      );
+    }
+    return new URL(
+      `${provider.apiBaseUrl}/projects/${owner}/repos/${name}/raw/${pathWithoutSlash}?at=${ref}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+// Converts for example
+// from: https://bitbucket.org/orgname/reponame/src/master/file.yaml
+// to:   https://api.bitbucket.org/2.0/repositories/orgname/reponame/src/master/file.yaml
+export function getRawUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+
+    if (
+      !owner ||
+      !name ||
+      (filepathtype !== 'browse' &&
+        filepathtype !== 'raw' &&
+        filepathtype !== 'src')
+    ) {
+      throw new Error('Invalid Bitbucket URL or file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+
+    if (provider.host === 'bitbucket.org') {
+      if (!ref) {
+        throw new Error('Invalid Bitbucket URL or file path');
+      }
+      return new URL(
+        `${provider.rawBaseUrl}/repositories/${owner}/${name}/src/${ref}/${pathWithoutSlash}`,
+      );
+    }
+    return new URL(
+      `${provider.rawBaseUrl}/projects/${owner}/repos/${name}/raw/${pathWithoutSlash}?at=${ref}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+export function readConfig(config: Config): ProviderConfig[] {
+  const providers: ProviderConfig[] = [];
 
   const providerConfigs =
     config.getOptionalConfigArray('integrations.bitbucket') ?? [];
 
+  // First read all the explicit providers
   for (const providerConfig of providerConfigs) {
     const host = providerConfig.getOptionalString('host') ?? 'bitbucket.org';
+    let apiBaseUrl = providerConfig.getOptionalString('apiBaseUrl');
+    let rawBaseUrl = providerConfig.getOptionalString('rawBaseUrl');
+    const token = providerConfig.getOptionalString('token');
+    const username = providerConfig.getOptionalString('username');
+    const password = providerConfig.getOptionalString('appPassword');
```

Comment:
here you call it `appPassword` instead of `password`, is there a particular reason?

Reply:
I would prefer `password`. However, the current bitbucket implementation defines the key `appPassword`. How would we proceed here? Do we want to take the opportunity to change it to password and deprecate the old property or should I stick with the already defined appPassword?

Reply:
Ah, I was not aware. Well I don't know the domain (for example what the bitbucket docs refers to it by), so I'd vote to just leave it as is for now unless there's reason to change

Reply:
App passwords iirc are a bit like access tokens for apps, i.e. you can create more of them for your account, but for some reason you need to include your username when using them too. I requested the `appPassword` naming so that people don't put their own password there by mistake.

Reply:
I kept the `appPassword` naming